File: gas_station_problem/java/src/SortSolution.java
import org.json.JSONArray;

import java.util.ArrayList;
import java.util.Collections;

public class SortSolution {

    public static ArrayList<Integer> getGasStations(JSONArray inputArr) {
        ArrayList<Integer> sortedArr = getSortedList(inputArr);
        ArrayList<Integer> outputList = new ArrayList<>();

        int currentMile = 0;
        int lastStation = 0;
        for(Integer station : sortedArr) {
            if (station - currentMile >= Constants.MILEAGE) {
                outputList.add(lastStation);
                currentMile = lastStation;
            } else {
                lastStation = station;
            }

            if (station == Constants.END_MILE) {
                outputList.add(station);
                break;
            }
        }

        return outputList;
    }

    private static ArrayList<Integer> getSortedList(JSONArray inputArr) {
        ArrayList<Integer> sorted = new ArrayList<>(inputArr.length());
        for (int i = 0; i < inputArr.length(); i++) {
            sorted.add(inputArr.optInt(i));
        }
        Collections.sort(sorted);
        return sorted;
    }


}

File: gas_station_problem/python/sort_solution.py
import sys
import json
import time

MILEAGE = 50
END_MILE = 10000

''' nlogn solution '''

def load_list(input_path):

    with open(input_path) as in_f:
        return json.load(in_f)


def get_gas_stations(mileage_list):

    sorted_list = sorted(mileage_list)
    out_list = []

    current_mile = 0
    last_station = 0
    for station in sorted_list:
        if station - current_mile >= MILEAGE:
            out_list.append(last_station)
            current_mile = last_station
        else:
            last_station = station

        if station == END_MILE:
            out_list.append(station)
            break
  
    return out_list


def main():
    if len(sys.argv) != 2:
        print('Incorrect number of args, requires input file')
        exit()

    input_list = load_list(sys.argv[1])
    
    start = time.time()
    output_solution = get_gas_stations(input_list)
    end = time.time()
    print(output_solution)
    print('Solution took:')
    print(end - start)


if __name__ == '__main__':
    main()

